plutus-tx/src/PlutusTx/Show.hs

```diff
@@ -20,28 +21,78 @@ import PlutusTx.Bool
 import PlutusTx.Builtins qualified as Builtins
 import PlutusTx.Either
 import PlutusTx.Foldable (foldl)
+import PlutusTx.List
 import PlutusTx.Maybe
 import PlutusTx.Show.TH
 
 instance Show Builtins.Integer where
-    {-# INLINEABLE show #-}
-    show = Builtins.showInteger
+    {-# INLINEABLE showsPrec #-}
+    showsPrec _ n =
+        showString (if n `Builtins.lessThanInteger` 0 then "-" else "")
+            . foldr alg id (toDigits n)
+      where
+        toDigits :: Builtins.Integer -> [Builtins.Integer]
+        toDigits m =
+            let quotient = m `Builtins.quotientInteger` 10
+                remainder = m `Builtins.remainderInteger` 10
+                digit =
+                    if remainder `Builtins.lessThanInteger` 0
+                        then 0 `Builtins.subtractInteger` remainder
+                        else remainder
+             in if quotient `Builtins.equalsInteger` 0
+                    then [digit]
+                    else digit : toDigits quotient
+        alg :: Builtins.Integer -> ShowS -> ShowS
+        alg digit acc =
+            acc
+                . showString
+                    ( if
+                        | digit `Builtins.equalsInteger` 0 -> "0"
+                        | digit `Builtins.equalsInteger` 1 -> "1"
+                        | digit `Builtins.equalsInteger` 2 -> "2"
+                        | digit `Builtins.equalsInteger` 3 -> "3"
+                        | digit `Builtins.equalsInteger` 4 -> "4"
+                        | digit `Builtins.equalsInteger` 5 -> "5"
+                        | digit `Builtins.equalsInteger` 6 -> "6"
+                        | digit `Builtins.equalsInteger` 7 -> "7"
+                        | digit `Builtins.equalsInteger` 8 -> "8"
+                        | digit `Builtins.equalsInteger` 9 -> "9"
+                        | otherwise                        -> "<invalid digit>"
+                    )
 
 instance Show Builtins.BuiltinByteString where
-    {-# INLINEABLE show #-}
-    show = Builtins.showByteString
-
-instance Show Builtins.BuiltinData where
-    {-# INLINEABLE show #-}
-    show = Builtins.showData
+    {-# INLINEABLE showsPrec #-}
+    -- Base16-encode the ByteString and show the result.
+    showsPrec _ s = foldl alg id [0 .. Builtins.subtractInteger (Builtins.lengthOfByteString s) 1]
```

Comment:
Why `foldl` here and `foldr` in `Show Builtins.Integer`? `foldr` is better for function composition.

Reply:
Can you elaborate why `foldr` is better for function composition? I didn't know it is. Is `f . (g . h)` better than `(f . g) . h` somehow?

Reply:
> Is `f . (g . h)` better than `(f . g) . h` somehow?

In Haskell, yes, the former reduces to `\x -> f ((g . h) x)`, while the latter to `\x -> (f . g) (h x)` and so the former can compute immediately if `f` doesn't force its argument while the latter has to reduce the inner composition first. Here's a test: the former is [able](https://ideone.com/UDYXdi) to handle an infinite list,  while the latter is [not](https://ideone.com/FWK7in).

Not that it's not only about infinite lists: you may want to get a `String`  lazily to print it as it gets computed, if the `String` is huge or it takes a lot of time to compute it fully right away.

In Plutus there's probably no difference due to the language being strict, though. So I guess `foldl` is equally fine.

I'd still prefer consistent usage of either `foldr` or `foldl`.

Reply:
> So I guess `foldl` is equally fine.

Although it may be slower or faster compared to `foldr`, hard to tell without benchmarking.

Reply:
Sounds good, thanks. Changed `foldl` to `foldr`.